Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp

```diff
@@ -0,0 +1,310 @@
+/*
+ * Copyright (C) 2022 Igalia S.L.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE COMPUTER, INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE COMPUTER, INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "NotificationService.h"
+
+#include "WebNotification.h"
+#include <gio/gio.h>
+#include <glib/gi18n-lib.h>
+#include <mutex>
+#include <wtf/FastMalloc.h>
+#include <wtf/Seconds.h>
+#include <wtf/UUID.h>
+#include <wtf/glib/GUniquePtr.h>
+#include <wtf/glib/Sandbox.h>
+#include <wtf/text/CString.h>
+
+#if PLATFORM(GTK)
+#include <WebCore/GtkVersioning.h>
+#endif
+
+namespace WebKit {
+
+static const Seconds s_dbusCallTimeout = 20_ms;
+
+NotificationService& NotificationService::singleton()
+{
+    static std::once_flag onceFlag;
+    static LazyNeverDestroyed<NotificationService> service;
+
+    std::call_once(onceFlag, [] {
+        service.construct();
+    });
+
+    return service;
+}
+
+NotificationService::NotificationService()
+{
+    const char* busName = shouldUsePortal() ? "org.freedesktop.portal.Desktop" : "org.freedesktop.Notifications";
+    const char* objectPath = shouldUsePortal() ? "/org/freedesktop/portal/desktop" : "/org/freedesktop/Notifications";
+    const char* interfaceName = shouldUsePortal() ? "org.freedesktop.portal.Notification" : "org.freedesktop.Notifications";
+    GUniqueOutPtr<GError> error;
+    m_proxy = g_dbus_proxy_new_for_bus_sync(G_BUS_TYPE_SESSION, G_DBUS_PROXY_FLAGS_DO_NOT_LOAD_PROPERTIES, nullptr, busName, objectPath, interfaceName, nullptr, &error.outPtr());
```

Comment:
Why do you have to use sync I/O here? Wouldn't this be quite easy to convert to async?

Reply:
Sync bus lookup is non-blocking after the first call.

Reply:
What's the "first call?"

Reply:
Using async IO for these would make everything a lot more complicated for very little benefit. For the first notification show will happen before the proxy is constructed or capabilities processed, so we would need to handle the case and wait for the async to complete to show the notification. Since this only happens once I don't think it's worth it.

Reply:
> What's the "first call?"

I think he means getting the bus connection, because g_bus_get_sync returns a singleton, so only the first time it's called it needs to connect which is async. At this point, glib has already connected to the session bus for sure.